astropy/utils/tests/test_data.py

```diff
@@ -430,62 +945,521 @@ def __init__(self, data):
+
+
+def test_clear_download_cache_refuses_to_delete_outside_the_cache(tmpdir):
+    fn = os.path.abspath(os.path.join(tmpdir, "file"))
+    with open(fn, "w") as f:
+        f.write("content")
+    assert os.path.exists(fn)
+    with pytest.raises(RuntimeError):
+        clear_download_cache(fn)
+    assert os.path.exists(fn)
+
+
+def test_check_download_cache_finds_unreferenced_files(temp_cache, valid_urls):
+    u, c = next(valid_urls)
+    download_file(u, cache=True)
+    with _cache(write=True) as (dldir, urlmap):
+        del urlmap[u]
+    with pytest.raises(ValueError):
+        check_download_cache()
+    clear_download_cache()
+
+
+def test_check_download_cache_finds_missing_files(temp_cache, valid_urls):
+    u, c = next(valid_urls)
+    os.remove(download_file(u, cache=True))
+    with pytest.raises(ValueError):
+        check_download_cache()
+    clear_download_cache()
+
+
+def test_check_download_cache_finds_bogus_entries(temp_cache, valid_urls):
+    u, c = next(valid_urls)
+    download_file(u, cache=True)
+    with _cache(write=True) as (dldir, urlmap):
+        bd = os.path.join(dldir, "bogus")
+        os.mkdir(bd)
+        bf = os.path.join(bd, "file")
+        with open(bf, "wt") as f:
+            f.write("bogus file that exists")
+        urlmap[u] = bf
+    with pytest.raises(ValueError):
+        check_download_cache()
+    clear_download_cache()
+
+
+def test_check_download_cache_finds_bogus_hashes(temp_cache, valid_urls):
+    u, c = next(valid_urls)
+    fn = download_file(u, cache=True)
+    with open(fn, "w") as f:
+        f.write("bogus contents")
+    with pytest.raises(ValueError):
+        check_download_cache(check_hashes=True)
+    clear_download_cache()
+
+
+def test_download_cache_update_doesnt_damage_cache(temp_cache, valid_urls):
+    u, _ = next(valid_urls)
+    download_file(u, cache=True)
+    download_file(u, cache="update")
+
+
+def test_cache_dir_is_actually_a_file(tmpdir, valid_urls):
+    """Ensure that bogus cache settings are handled sensibly.
+
+    Because the user can specify the cache location in a config file, and
+    because they might try to deduce the location by looking around at what's
+    in their directory tree, and because the cache directory is actual several
+    tree levels down from the directory set in the config file, it's important
+    to check what happens if each of the steps in the path is wrong somehow.
+    """
+    def check_quietly_ignores_bogus_cache():
+        """We want a broken cache to produce a warning but then astropy should
+        act like there isn't a cache.
+        """
+        with pytest.warns(CacheMissingWarning) as w:
+            assert not get_cached_urls()
+        assert any(wi.category == CacheMissingWarning for wi in w)
+        with pytest.warns(CacheMissingWarning) as w:
+            assert not is_url_in_cache("http://www.example.com/")
+        assert any(wi.category == CacheMissingWarning for wi in w)
```

Comment:
I am not very sure why `assert any(wi.category == CacheMissingWarning for wi in w)` is necessary. If `pytest.warns` detects that the warning is not emitted, the test would fail. But if it passes `pytest.warns`, wouldn't that `assert` always be `True` anyway? Am I missing something?